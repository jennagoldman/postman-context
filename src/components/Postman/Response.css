.responseSection {
  font-size: 1em;
  max-height: 600px;
  overflow: auto;
  width: 55rem;
}