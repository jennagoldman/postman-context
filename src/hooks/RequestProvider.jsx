import React, { createContext, useState, useEffect, useContext } from 'react';
import makeRequest from '../services/make-request.js';
import PropTypes from 'prop-types';
import RequestForm from '../components/Postman/RequestForm.jsx';

export const RequestContext = createContext();

export const RequestProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('');
  const [body, setBody] = useState('');
  const [response, setResponse] = useState('');
  const [history, setHistory] = useState([]);

  const handleChange = ({ target }) => {
    if(target.name === 'url') setUrl(target.value);
    if(target.name === 'method') setMethod(target.value);
    if(target.name === 'body') setBody(target.value);
  }

  useEffect(() => {
    if(localStorage.getItem('requestHistory') === null) {
      localStorage.setItem('requestHistory', JSON.stringify([]));
    }
    const history = localStorage.getItem('requestHistory');
    setHistory(JSON.parse(history));
  }, []);

  useEffect(() => {
    localStorage.setItem('requestHistory', JSON.stringify(history));
  }, [history]);

  const handleRequestSubmit = (event) => {
    event.preventDefault();
    makeRequest(url, method, body)
      .then(response => setResponse(response));

    setHistory(prevHistory => ([
      ...prevHistory,
      {
        url,
        method,
        body
      }
    ]));
    localStorage.setItem('requestHistory', JSON.stringify(history));
  };

  const handleHistoryReset = () => {
    localStorage.setItem('requestHistory', JSON.stringify([]));
    setHistory([]);
  };

  return (
    <RequestContext.Provider value={{ url, method, body, response, history, handleChange, handleRequestSubmit, handleHistoryReset}}>
      {children}
    </RequestContext.Provider>
  )
}

RequestProvider.propTypes = {
  children: PropTypes.node
}

export const useRequestContext = () => {
  const context = useContext(RequestContext);
  return context;
}

export const useResponse = () => {
  const { response } = useContext(RequestContext);
  return response;
}