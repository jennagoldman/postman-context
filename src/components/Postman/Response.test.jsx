  
import React from 'react';
import { shallow } from 'enzyme';
import { RequestProvider } from '../../hooks/RequestProvider.jsx';
import Response from './Response.jsx';

describe('Response', () => {
  it('matches a snapshot', () => {
    const wrapper = shallow(
      <RequestProvider>
        <Response />
      </RequestProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});