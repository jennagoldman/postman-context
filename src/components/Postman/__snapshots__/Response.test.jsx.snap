// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleChange": [Function],
      "handleHistoryReset": [Function],
      "handleRequestSubmit": [Function],
      "history": Array [],
      "method": "",
      "response": "",
      "url": "",
    }
  }
>
  <Response />
</ContextProvider>
`;
